tools/setup_services.py

```diff
@@ -1,15 +1,396 @@
+        requirement_string (str): The requirement specification (e.g., ">=1.0.0", "==1.2.3", ">=1.0,<2.0")
+
+    Returns:
+        bool: True if version satisfies the requirement, False otherwise
+    """
+    try:
+        v = version.parse(version_string)
+        spec = SpecifierSet(requirement_string)
+        return v in spec
+    except Exception as e:
+        log(f"Error validating version: {e}")
+        return False
+
+
+def validate_plugin_compatibility(plugin_name: str, plugin_path: str):
+    plugin_config = read_plugin_config(plugin_path)
+    name, version_requirement = get_plugin_compatible_version(plugin_config)
+    if not (name and version_requirement):
+        return True
+    try:
+        mod = __import__(name)
+        installed_version = mod.__version__
+
+        if not validate_version(installed_version, version_requirement):
+            log(
+                f"Plugin {plugin_name} is incompatible: required {version_requirement}, found {installed_version}. Skipping."
+            )
+            return False
+    except ImportError:
+        log(
+            f"Could not find package {name} which plugin {plugin_name} depends on. Skipping."
+        )
+        return False
+
+    if plugin_config["project"]["name"] != plugin_name:
+        log(
+            f"Plugin {plugin_name} has a different name in its config ({plugin_config['project']['name']}). Skipping."
+        )
+        return False
+
+    return True
+
+
+def run_git_command(args, plugin_path, plugin_name):
+    """
+    Run a git command in the specified plugin path.
+
+    Args:
+        args (list): Git command, e.g. ['checkout', 'main']
+        plugin_path (str): Directory where the command runs
+        plugin_name (str): Used in error logs
+
+    Returns:
+        stdout_lines (list of str): Output lines from stdout
+        stderr_str (str): Full stderr output
+
+    Raises:
+        subprocess.CalledProcessError: If the command fails
+    """
+    try:
+        result = subprocess.run(
+            ["git"] + args,
+            cwd=plugin_path,
+            capture_output=True,
+            text=True,
+            check=True,
+        )
+
+        return result.stdout.splitlines(), result.stderr
+
+    except subprocess.CalledProcessError as e:
+        msg = f"[ERROR] Git command failed: {' '.join(args)}"
+        if plugin_name:
+            msg += f" (plugin: {plugin_name})"
+        msg += f"\n{e.stderr}"
+        print(msg)
+        raise
+
+
+def download_plugin_services():
+    _, cfg = load_env()
+    plugins_path = cfg.get("services.plugins_path", "./plugins")
+    os.makedirs(plugins_path, exist_ok=True)
+
+    plugins = cfg.get("plugins", {})
+    plugin_services = []
+
+    log(f"Discovered {len(plugins)} plugins")
+
+    for plugin_name, plugin_info in plugins.items():
+        if "url" not in plugin_info:
+            log(f"Skipping plugin {plugin_name} because it has no URL")
```

Comment:
```suggestion
            log(f"Skipping plugin {plugin_name}: no URL found in config")
```

Maybe provide some guidance here even, as to where and how user should specify config.

Reply:
still a TODO

Reply:
well examples are given in the config default, so maybe that's enough?